src/responder/multiResponder.ts

```diff
@@ -0,0 +1,237 @@
+        try {
+            // since we've made this method available publicly we need to ensure that the class has been initialised
+            await this.setup();
+
+            if (this.queue.queueItems.length === 0) {
+                throw new QueueConsistencyError(
+                    `Transaction mined for empty queue at nonce ${nonce}. ${inspect(txIdenfifier)}`
+                );
+            }
+            if (!this.queue.contains(txIdenfifier)) {
+                throw new QueueConsistencyError(`Transaction identifier not found in queue. ${inspect(txIdenfifier)}`);
+            }
+            const frontItem = this.queue.queueItems[0];
+            if (frontItem.nonce !== nonce) {
+                throw new QueueConsistencyError(
+                    `Front of queue nonce ${frontItem.nonce} does not correspond to nonce ${nonce}. ${inspect(
+                        txIdenfifier
+                    )}`
+                );
+            }
+
+            if (txIdenfifier.equals(frontItem.request.identifier)) {
+                // the mined transaction was the one at the front of the current queue
+                // this is what we hoped for, simply dequeue the transaction
+                this.queue = this.queue.dequeue();
+            } else {
+                // the mined transaction was not the one at the front of the current queue
+                // - it was at the front of a past queue. This means that the transaction
+                // at the front of the current queue can no longer be mined as it shares the same
+                // nonce. We need to find the transaction in the current queue that corresponds to
+                // the mined tx and remove it. In doing so free up a later nonce.
+                // and bump up all transactions with a lower nonce so that the tx that is
+                // at the front of the current queue - but was not mined - remains there
+                const reducedQueue = this.queue.consume(txIdenfifier);
+                const replacedTransactions = reducedQueue.difference(this.queue);
+                this.queue = reducedQueue;
+
+                // since we had to bump up some transactions - change their nonces
+                // we'll have to issue new transactions to the network
+                await Promise.all(replacedTransactions.map(this.broadcast));
+            }
+        } catch (doh) {
+            if (doh instanceof QueueConsistencyError) logger.error(doh.stack!);
+            else {
+                logger.error(`Unexpected error after mining transaction. ${txIdenfifier}.`);
+                if (doh.stack) logger.error(doh.stack);
+                else logger.error(doh);
+            }
+        }
+    }
+
+    private async broadcast(queueItem: GasQueueItem) {
+        try {
+            this.transactionTracker.addTx(queueItem.request.identifier, this.txMined);
+            await this.signer.sendTransaction(queueItem.toTransactionRequest());
```

Comment:
Hmm, I'm not sure about this one. I think it's fine for the queue to mutate whilst broadcast is occurring. What kind of problems do you see happening if this does occur?

Reply:
I'm always wary of collections being changed while you have an iterator, as it is hard to be convinced that there is no bug. Here, disaster would be not broadcasting a transaction. Can it happen? Probably not, but it is hard to be absolutely sure. In mission critical places, I would prefer to avoid this (potential) source of problems.

Perhaps we can just suspend the previous broadcast if we are starting a new one? E.g.: have a private broadcastNumber member in the class incremented at each broadcast, and abort if it is incremented before we finish the broadcast? (problem with this: we need to make sure that broadcasts don't keep getting suspended, or we never reach the end!)

Reply:
I think it's fine after discussing it.